TR123/Autosplitter.cs

```diff
@@ -0,0 +1,256 @@
+    /// <param name="state"><see cref="LiveSplitState" /> passed by LiveSplit</param>
+    /// <returns><see langword="true" /> if the timer should split, <see langword="false" /> otherwise</returns>
+    public bool ShouldSplit(LiveSplitState state)
+    {
+        // Determine if the player is in a level we have not already split.
+        uint currentLevel = GameData.CurrentLevel();
+        bool onValidLevelToSplit = AllGameStats[CurrentActiveGame].LevelStats.All(stats => stats.LevelNumber != currentLevel);
+        if (!onValidLevelToSplit)
+            return false;
+
+        // Deathrun
+        var health = GameData.Health;
+        if (Settings.Deathrun)
+        {
+            bool laraJustDied = health.Old > 0 && health.Current <= 0;
+            return laraJustDied;
+        }
+
+        // FG & IL/Section
+        var levelComplete = GameData.LevelComplete;
+        bool levelJustCompleted = !levelComplete.Old && levelComplete.Current;
+        return levelJustCompleted;
+    }
+
+    /// <summary>Determines if the timer should reset.</summary>
+    /// <param name="state"><see cref="LiveSplitState" /> passed by LiveSplit</param>
+    /// <returns><see langword="true" /> if the timer should reset, <see langword="false" /> otherwise</returns>
+    public bool ShouldReset(LiveSplitState state)
+    {
+        if (!Settings.EnableAutoReset)
+            return false;
+
+        var titleLoaded = GameData.TitleLoaded;
+        bool justEnteredTitleScreen = !titleLoaded.Old && titleLoaded.Current;
+        if (!justEnteredTitleScreen)
+            return false;
+
+        // Checking LevelComplete is inaccurate; depending on exactly when LiveSplit performs its polling, the Old and Current values may have display different values.
+        uint level = GameData.CurrentLevel();
+        switch (CurrentActiveGame)
+        {
+            case Game.Tr1 or Game.Tr1NgPlus or Game.Tr1UnfinishedBusiness:
+            {
+                var levelCutscene = GameData.Tr1LevelCutscene;
+                bool wasPossiblyOnLastLevel =
+                    levelCutscene.Old is (uint)Tr1Level.TheGreatPyramid or (uint)Tr1Level.TempleOfTheCat;
+                bool oldLevelWasOnCreditsValue = level == 1;
+                return !wasPossiblyOnLastLevel || !oldLevelWasOnCreditsValue;
+            }
+            case Game.Tr2 or Game.Tr2NgPlus or Game.Tr2GoldenMask:
+            {
+                bool wasPossiblyOnLastLevel =
+                    level is (uint)Tr2Level.HomeSweetHome or (uint)Tr2Level.Kingdom or (uint)Tr2Level.NightmareInVegas;
+                return !wasPossiblyOnLastLevel;
+            }
+            case Game.Tr3 or Game.Tr3NgPlus or Game.Tr3TheLostArtifact:
+            {
+                bool wasPossiblyOnLastLevel =
+                    level is (uint)Tr3Level.MeteoriteCavern or (uint)Tr3Level.AllHallows or (uint)Tr3Level.Reunion;
+                return !wasPossiblyOnLastLevel;
+            }
```

Comment:
I'd cast `level` to the appropriate game enum so you don't have to cast the level values to uint several times.